ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/bytes/VarInt.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2014-2023 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.quic.bytes
+
+import io.ktor.network.quic.util.*
+import io.ktor.utils.io.core.*
+
+/**
+ * Reads variable-length non-negative integer value. Returns null if EOF.
+ *
+ * [RFC Reference](https://www.rfc-editor.org/rfc/rfc9000.html#name-variable-length-integer-enc)
+ */
+@OptIn(ExperimentalUnsignedTypes::class)
+internal fun ByteReadPacket.readVarIntOrNull(): Long? {
```

Comment:
I tried to do it with value class in the first place, but it seemed redundant, as it is only used in the decoding and encoding of the frames. It this particular case there can be -1 as the `null` value, but maybe the approach as with  `readUByteOrIfEmpty` be even better

Reply:
but i don't want to make it inline, so maybe something like this would work, wdyt?
```
internal inline fun ByteReadPacket.readVarIntOrElse(elseBlock: () -> Nothing): Long {
    val value = readVarInt()
    if (value == -1L) {
        elseBlock()
    }
    return value
}

internal fun ByteReadPacket.readVarInt(): Long {
    val value = readUByteOrElse { return -1 }.toInt()
    ...
```

Reply:
Yes, that's the right approach IMO (considering how it's used in code), `private no-inline` function, which returns -1 and `internal inline` which check for `null`